### ISSUE TITLE
PanicException after trying to take one row from a grouped dataset with pl.List(Date) column

### ISSUE THREAD

**Issue by jakubkovac** [pola-rs/polars#11490]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

EDIT: works in `0.19.3`, does not work in later versions including `0.19.7`
EDIT: attaching a zipped parquet file that reproduces the issue [df.zip](https://github.com/pola-rs/polars/files/12819927/df.zip)

### Reproducible example

This will be very hard to reproduce because of the issue I'm experiencing. Let's say I have a dataframe `df`.
The issue is that if I read the data from the output of `df.to_init_repr()` while changing the dtype of the column b from `pl.List(Date)` to `pl.List(pl.Date)` the issue is no longer present.

```python
df
df.schema
df.to_init_repr()
df2 = df.with_columns(pl.col("b").cast(pl.List(pl.Date)))
df2.schema
df2.to_init_repr()
df2.group_by("a").head(1)
df.group_by("a").head(1)
```

### Log output

```shell
shape: (2, 2)
┌─────┬───────────────────────────────────┐
│ a   ┆ b                                 │
│ --- ┆ ---                               │
│ i32 ┆ list[date]                        │
╞═════╪═══════════════════════════════════╡
│ 1   ┆ [2024-03-01, 2024-03-01, 2024-03… │
│ 1   ┆ [2024-07-01, 2024-07-01]          │
└─────┴───────────────────────────────────┘

{'a': Int32, 'b': List(Date)}

pl.DataFrame(
    [
        pl.Series("a", [1, 1], dtype=pl.Int32),
        pl.Series("b", [[datetime.date(2024, 3, 1), datetime.date(2024, 3, 1), datetime.date(2024, 3, 1)], [datetime.date(2024, 7, 1), datetime.date(2024, 7, 1)]], dtype=pl.List(Date)),
    ]
)

{'a': Int32, 'b': List(Date)}

pl.DataFrame(
    [
        pl.Series("a", [1, 1], dtype=pl.Int32),
        pl.Series("b", [[datetime.date(2024, 3, 1), datetime.date(2024, 3, 1), datetime.date(2024, 3, 1)], [datetime.date(2024, 7, 1), datetime.date(2024, 7, 1)]], dtype=pl.List(Date)),
    ]
)

shape: (1, 2)
┌─────┬───────────────────────────────────┐
│ a   ┆ b                                 │
│ --- ┆ ---                               │
│ i32 ┆ list[date]                        │
╞═════╪═══════════════════════════════════╡
│ 1   ┆ [2024-03-01, 2024-03-01, 2024-03… │
└─────┴───────────────────────────────────┘

---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
File <command-4418020169336725>:11
      8 print(df2.to_init_repr())
     10 print(df2.group_by("a").head(1))
---> 11 print(df.group_by("a").head(1))

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/polars/dataframe/group_by.py:374, in GroupBy.head(self, n)
    327 def head(self, n: int = 5) -> DataFrame:
    328     """
    329     Get the first `n` rows of each group.
    330 
   (...)
    371 
    372     """
    373     return (
--> 374         self.df.lazy()
    375         .group_by(self.by, *self.more_by, maintain_order=self.maintain_order)
    376         .head(n)
    377         .collect(no_optimization=True)
    378     )

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/polars/utils/deprecation.py:95, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     90 @wraps(function)
     91 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     92     _rename_keyword_argument(
     93         old_name, new_name, kwargs, function.__name__, version
     94     )
---> 95     return function(*args, **kwargs)

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/polars/lazyframe/frame.py:1713, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, no_optimization, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, streaming, **kwargs)
   1700     comm_subplan_elim = False
   1702 ldf = self._ldf.optimization_toggle(
   1703     type_coercion,
   1704     predicate_pushdown,
   (...)
   1711     eager,
   1712 )
-> 1713 return wrap_df(ldf.collect())

PanicException: called `Result::unwrap()` on an `Err` value: OutOfSpec("ListArray's child's DataType must match. However, the expected DataType is Int32 while it got Date32.")
```

### Issue description

The output of `df.to_init_repr()` is identical to `df2.to_init_repr()` , the schema also says the same thing.

### Expected behavior

 I would expect that `group_by("a").head(1)` would work on both dataframes.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.6
Index type:          UInt32
Platform:            Linux-5.15.0-1042-azure-x86_64-with-glibc2.35
Python:              3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.0.0
connectorx:          <not installed>
deltalake:           0.11.0
fsspec:              2022.7.1
gevent:              <not installed>
matplotlib:          3.5.2
numpy:               1.21.5
openpyxl:            3.1.2
pandas:              1.4.4
pyarrow:             12.0.0
pydantic:            1.10.7
pyiceberg:           <not installed>
pyxlsb:              <not installed>
sqlalchemy:          2.0.21
xlsx2csv:            <not installed>
xlsxwriter:          3.1.6
```

</details>

**Comment by cmdlineluser** [comment]
Are you saying that the following is raising the error for you?
```python
import datetime
import polars as pl

df = pl.DataFrame(
    [
        pl.Series("a", [1, 1], dtype=pl.Int32),
        pl.Series("b", [[datetime.date(2024, 3, 1), datetime.date(2024, 3, 1), datetime.date(2024, 3, 1)], [datetime.date(2024, 7, 1), datetime.date(2024, 7, 1)]], dtype=pl.List(pl.Date)),
    ]
)

df.group_by("a").head(1)

# shape: (1, 2)
# ┌─────┬───────────────────────────────────┐
# │ a   ┆ b                                 │
# │ --- ┆ ---                               │
# │ i32 ┆ list[date]                        │
# ╞═════╪═══════════════════════════════════╡
# │ 1   ┆ [2024-03-01, 2024-03-01, 2024-03… │
# └─────┴───────────────────────────────────┘
```

**Comment by jakubkovac** [comment]
No, the code you wrote actually works. The problem is I have a dataframe that has that exact same representation(when using the method df.to_init_repr()) but the `df.group_by("a").head(1)` does not work on it and raises the exception I pasted above.
I cannot share the exact process of how I got that dataframe but I got it by running a `group_by("another_column").agg(pl.col("b").alias("b"))`

**Comment by sahuagin** [comment]
I've just run into this issue (again) and wanted to leave my workaround for other should they arrive here.
While I tried to create a MVE, I was unable to reproduce outside of the production testing.  In our case, we have a more complicated setup that most, and are creating a Vec<Vec<NaiveDate>> in rust which ends up being Series<ChunkedDate> per row, and then using pyo3-polars, we are returning this dataframe into python, where a number of operations are done on the dataframe successfully, but at the end of the test the dataframe is sorted by 2 different columns (than the chunkedDate column) which causes the type mismatch.

tl;dr - in python, just recast the column to it's type and the assertion goes away.
```python
df = df.lazy().with_columns([pl.col("list_o_dates").cast(pl.List(pl.Date))]).collect()
df.sort(by=[other0, other1])
# success
```
